core/controllers/async_download.py

```diff
@@ -0,0 +1,147 @@
+import io
+from datetime import datetime
+from typing import IO
+
+from flask import current_app
+from notifications_python_client.notifications import NotificationsAPIClient
+from werkzeug.datastructures import FileStorage
+
+from config import Config
+from core.aws import _S3_CLIENT
+from core.controllers.download import download
+
+
+def async_download(
+    email_address: str,
+    file_format: str,
+    funds: list[str] | None = None,
+    organisations: list[str] | None = None,
+    regions: list[str] | None = None,
+    rp_start: str | None = None,
+    rp_end: str | None = None,
+    outcome_categories: list[str] | None = None,
+):
+    """Download data, store file in S3 and send an email to the user with the download link."""
+
+    response = download(
+        file_format=file_format,
+        funds=funds,
+        organisations=organisations,
+        regions=regions,
+        rp_start=rp_start,
+        rp_end=rp_end,
+        outcome_categories=outcome_categories,
+    )
+
+    if response.status_code != 200:
+        current_app.logger.error(
+            "Error downloading data from data-store/download endpoint. Status code: {response_status}",
+            extra=dict(response_status=response.status_code),
+        )
+        return
+
+    file_format_str: str = get_file_format_from_extension(file_format)
+    file_size_str: str = get_human_readable_file_size(len(response.data))
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+    email_template_id = Config.NOTIFY_FIND_REPORT_DOWNLOAD_TEMPLATE_ID
+
+    if not bucket:
+        current_app.logger.error("AWS_S3_BUCKET_FIND_DATA_FILES is not set.")
+    elif not find_service_base_url:
+        current_app.logger.error("FIND_SERVICE_BASE_URL is not set.")
+    elif not email_template_id:
+        current_app.logger.error("NOTIFY_TEMPLATE_ID is not set.")
```

Comment:
This is returning a 204 from the endpoint, indicating success. This is fine for now since no (end) users will see it, and we're about to turn this into a background job, but we'll need to consider these error states more thoroughly when this is running on the queue. It should probably fail the task and raise an error to be handled fully rather than just logging an error message.

Reply:
Removed these logging lines.